Consolidate Service Account classes.

dhermes opened the pull request:
Attempt at #211 (and integrating changes from #1 in as well).

@nathanielmanistaatgoogle This doesn't implement the scope-less credentials yet but I wanted you to get a chance to look at it before then. I am going to remove the `SignedJwtAssertionCredentials` class and will probably update this PR when that is done as well.

dhermes committed:
Making service_account._ServiceAccountCredentials public.

dhermes committed:
Adding from_json_keyfile factory to ServiceAccountCredentials.

dhermes committed:
Making scopes optional in ServiceAccountCredentials.

dhermes committed:
Adding Pure Python Crypto module.

dhermes committed:
Dropping custom crypto in service_account.py.

Using the crypt module as the rest of the library does.

dhermes committed:
Making private_key_id and service_account_id optional in ServiceAccountCredentials.

dhermes committed:
Making ServiceAccount support .p12 credentials.

nathanielmanistaatgoogle commented on oauth2client/_pure_python_crypt.py:
```diff
@@ -0,0 +1,182 @@
+# limitations under the License.
+
+"""Pure python crypto-related routines for oauth2client."""
+
```
Only one blank line between module doc string and imports.

dhermes replied:
Changing it, but don't actually know what the rule is. (This is one of these, "I don't know what to do with my hands" situations.)

nathanielmanistaatgoogle replied:
Always two blank lines before any top-level `class` or `def` and never two blank lines anywhere else ever: https://google.github.io/styleguide/pyguide.html?showone=Blank_Lines#Blank_Lines.

dhermes commented on oauth2client/_pure_python_crypt.py:
```diff
@@ -0,0 +1,182 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Pure python crypto-related routines for oauth2client."""
```
I should capitalize Python and give more details how this is made possible. Though this module will never end up in Sphinx docs since it is private.

nathanielmanistaatgoogle commented on oauth2client/_pure_python_crypt.py:
```diff
@@ -0,0 +1,182 @@
+        except (ValueError, rsa.pkcs1.VerificationError):
+            return False
+
+    @staticmethod
```
Since this is brand new code in a brand new module, can we eschew the use of `staticmethod` and have this just be a module-scope construction function?

dhermes replied:
`RsaVerifier` needs to be able to stand in as `oauth2client.crypt.Verifier`. This object is used throughout and our code (and user code) should be able to call ``-oauth2client.crypt.Verifier.from_string` and get a verifier whether they have pyOpenSSL, PyCrypto, or just the regular dependencies installed.

nathanielmanistaatgoogle replied:
In that case it's worth an implementation comment (`#`) explaining that this class object must have the same class interface (static methods and all) as `oauth2client.crypt.Verifier`.

Now a larger question (probably outside the scope of this commit) is: what freedom do we have to change the interface of `oauth2client.crypt.Verifier`? It's never been documented as a part of the library's public API, has it? Is it one of the many things that was only ever public because someone didn't think to prepend an underscore to its name? Do we think it's in use by users of the library?

dhermes replied:
It should just be a `classmethod`, as should all the others. It's possible the original implementers didn't know how to make a class method?

nathanielmanistaatgoogle commented on oauth2client/_pure_python_crypt.py:
```diff
@@ -0,0 +1,182 @@
+        pkey: rsa.key.PrivateKey (or equiv), The private key to sign with.
+    """
+
+    _PKCS1_MARKER = ('-----BEGIN RSA PRIVATE KEY-----',
```
Can these constants be pulled out of the class and placed at module scope?

dhermes replied:
Happy to do it, though the further ~~the~~ they get from `RsaSigner.from_string`, the more the context is lost. (In other words, I put them there on purpose.)

nathanielmanistaatgoogle replied:
A change that reduces the number of code elements inside a class is almost always a good trade-off for the overall codebase because classes are the most inherently complex language construct. A slight loss of textual-spatial locality is worth it here to communicate to the future maintainer that these values aren't used via the class object, shadowed by subclasses, accessed by users of the class, or anything else that would be an open question if the values were located within the class.

dhermes replied:
Doing it now.

nathanielmanistaatgoogle commented on oauth2client/_pure_python_crypt.py:
```diff
@@ -0,0 +1,182 @@
+        message = _to_bytes(message, encoding='utf-8')
+        return rsa.pkcs1.sign(message, self._key, 'SHA-256')
+
+    @classmethod
```
If the constants are pulled out to module scope, can this construction function also be extracted?

dhermes replied:
Ditto here as above. `RsaSigner` needs to stand in as `oauth2client.crypt.Signer`.

nathanielmanistaatgoogle replied:
Definitely worth a `#`-delimited implementation comment explaining that this code only looks this way to satisfy the larger constraint of "this class must present the same class interface as that other class".

And the same outside-the-scope-of-this-change question is raised asking "how much freedom do we have to change this interface?".

dhermes replied:
Is there something wrong with the interface?

nathanielmanistaatgoogle replied:
I personally don't like class methods, especially class methods used for construction, but that's not what's relevant here.

What's relevant here is that if this class is intended to be used in place of another class then it must maintain the same public API as that other class. If a method is added over there, the same method must be added here. If someone wonders why the methods of this class have the signature that they have, they should look over there. Right now that's not obvious from looking at this class - it doesn't inherit from the class that controls its interface design (although I'm not saying it should) nor does it reference that class in any other way. One wouldn't know from this class' text that it has any relationship to any other class, right? That's what strikes me as a maintenance danger that an implementation comment should address.

Does the instance constructor also have to have the same signature as the instance constructor of the other class?

nathanielmanistaatgoogle commented on oauth2client/service_account.py:
```diff
@@ -12,96 +12,238 @@
+                        ``private_key_password`` is not passed in.
+    """
+
+    MAX_TOKEN_LIFETIME_SECS = 3600
```
Is it intended that this be a public attribute of the class? It looks like a plain and simple boring constant?

dhermes replied:
It's used in `_generate_assertion`, with object-level overrides allowed.

nathanielmanistaatgoogle commented on tests/test__pure_python_crypt.py:
```diff
@@ -0,0 +1,178 @@
+# limitations under the License.
+
+"""Unit tests for oauth2client._pure_python_crypt."""
+
```
Single blank line here too.

nathanielmanistaatgoogle commented on tests/test__pure_python_crypt.py:
```diff
@@ -0,0 +1,178 @@
+    def test_from_string_pub_key(self):
+        public_key = self._load_public_key_bytes()
+        verifier = RsaVerifier.from_string(public_key, is_x509_cert=False)
+        self.assertTrue(isinstance(verifier, RsaVerifier))
```
`assertIsInstance`, here and throughout everything working with `unittest2`.

nathanielmanistaatgoogle commented on tests/test__pycrypto_crypt.py:
```diff
@@ -23,7 +23,7 @@
 class TestPyCryptoVerifier(unittest.TestCase):
 
     PUBLIC_KEY_FILENAME = os.path.join(os.path.dirname(__file__),
```
Does the name of this constant need to change the way the name of the file is changing?

nathanielmanistaatgoogle commented:
Looks mostly really good.

dhermes committed:
Removing SignedJWTAssertionCredentials.

dhermes committed:
Addressing review comments.

dhermes commented:
@nathanielmanistaatgoogle Addressed review comments (with some remarks) and also added a commit to remove the old service account implementation (as well as to use `json_str` instead of `s` in all of the `from_json` methods)

dhermes commented:
@nathanielmanistaatgoogle FYI https://bitbucket.org/ned/coveragepy/issues/473/context-manager-confuses-branch-miss is causing `test_service_account.py` to report less than 100% coverage but it'll resolve itself once the update mentioned in the Bitbucket issue is released

dhermes commented:
@nathanielmanistaatgoogle It'd be nice to use `ServiceAccountCredentials.from_json_keyfile` in `GoogleCredentials.from_json` and `oauth2client.client._get_application_default_credential_from_file`, but as of right now the factory only takes a filename. What do you think about making `from_json_keyfile` accept one of three objects as its first argument:

- filename
- file / stream object
- parsed dictionary (that used to be JSON)

nathanielmanistaatgoogle commented on oauth2client/service_account.py:
```diff
@@ -12,96 +12,265 @@
+    Args:
+        service_account_email: string, The email associated with the
+                               service account.
+        private_key_pkcs8_pem: bytes, The content of a PKCS#8 key in PEM
```
Is this optional? It's associated with a keyword argument.

dhermes replied:
Yes. The JSON key files have PKCS#8 keys stored in PEM format (as a field in a JSON dict) and the .p12 key files have PKCS#12 keys/certs stored in DER format.

nathanielmanistaatgoogle replied:
That's what I thought - document it as "(Optional)" the same way the other optional parameters are documented?

dhermes replied:
Good call. Accidental miss.

nathanielmanistaatgoogle commented on oauth2client/service_account.py:
```diff
@@ -12,96 +12,265 @@
+                    access token.
+        _revoke_uri: string, (Optional) The URI to use when revoking an
+                     access token.
+        kwargs: dict, Extra key-value pairs to send in the payload body
```
Do these keys and values have to be of any particular type?

dhermes replied:
Probably string for both. They are snuck into the assertion payload (e.g. sending `sub` for subject is in our unit tests).

nathanielmanistaatgoogle replied:
Well, document that fact? Always be reticent to describe the type of something as "sequence" or "iterable" or "dict" instead of describing it as "list of some type" or "iterable of some type" or "dict from some key type to some value type".

dhermes replied:
Done. `kwargs` is always a `dict`, since `**kwargs` comes from the runtime.

nathanielmanistaatgoogle commented on oauth2client/service_account.py:
```diff
@@ -12,96 +12,265 @@
+        self._private_key_pkcs8_pem = private_key_pkcs8_pem
+        self._private_key_pkcs12 = private_key_pkcs12
+        self._private_key_password = private_key_password
+        self._signer = self._get_signer()
```
Never call an instance method from the constructor of the instance. This method doesn't seem to mutate the instance at all; why not just have it be a module-scope helper function that accepts three or four parameters?

dhermes replied:
Doing now.

dhermes replied:
It's funny I recently re-factored a research project a ton because the constructors were so hard to test/stub-out (since I used lots of instance methods in the constructor).

nathanielmanistaatgoogle commented on oauth2client/service_account.py:
```diff
@@ -12,96 +12,265 @@
-class _ServiceAccountCredentials(AssertionCredentials):
-    """Class representing a service account (signed JWT) credential."""
+        Raises:
+            ValueError: If both ``private_key_pkcs8_pem`` and
```
Never promise a certain response to a programming defect. In cases like this, simply say "it is illegal to pass non-`None` values for both `<X>` and `<Y>` in the same call" or something like that. Leave it to them to wonder what will happen if they violate the API - will a `ValueError` be raised? Will the program segfault? Will kittens be slaughtered? Give yourself the freedom to do any of those and take away from the users the freedom to rely on any particular specific punishment for a specific programming transgression.

dhermes replied:
I think we may differ here. This is the whole point of a `Raises` section. Types and all.

nathanielmanistaatgoogle replied:
Leaving the "Raises: ValueError: ..." is okay; for now we can agree to disagree. Next time we're away from code and just talking software philosophy over a meal I'll try to persuade you otherwise.

nathanielmanistaatgoogle commented on oauth2client/service_account.py:
```diff
@@ -12,96 +12,265 @@
+                            will be ``foo.apps.googleusercontent.com``.
+        user_agent: string, (Optional) User agent to use when sending
+                    request.
+        _token_uri: string, (Optional) The URI to use when requesting an
```
Underscore-prefixed parameters? Do we have to?

dhermes replied:
No we don't have to, I just did because when you and I discussed it, we agreed that no one should ever change these values.

nathanielmanistaatgoogle replied:
In that case... why are they even parameters? What's the advantage of a parameter that must never be set as opposed to an internal constant?

nathanielmanistaatgoogle commented on oauth2client/service_account.py:
```diff
@@ -12,96 +12,265 @@
+                        ``private_key_pkcs12`` are set.
+            ValueError: If ``private_key_pkcs8_pem`` is set and
+                        ``private_key_password`` is passed in.
+            ValueError: If ``private_key_pkcs11`` is set and
```
This constructor does not have a `private_key_pkcs11` parameter. Right? Typo?

dhermes replied:
Fixed. Good eye

nathanielmanistaatgoogle commented on oauth2client/service_account.py:
```diff
@@ -12,96 +12,265 @@
 
-class _ServiceAccountCredentials(AssertionCredentials):
-    """Class representing a service account (signed JWT) credential."""
+        Raises:
```
Here we risk creating a monster, by which I particularly mean "a public code element of our library the incorrect use of which is straightforwardly easy and the correct use of which is difficult". In a shiny beautiful perfect world, an invokable code element (class constructor, method, function, or... one of a few other things) has parameters each of which can vary across the full range of their types and none of which are interdependent with any others. Now we don't always get to live in that shiny beautiful perfect world, but we are responsible for trying to avoid creating invokable elements in which there are too many ways to go wrong.

After taking a little while longer to think about this... this seems like a globally unattractive local optimum. If we constrain ourselves to have one public class for a "C", and we're in a situation in which a "C" can be constructed from an "X", or from a "Y", or from a "Z", then this is just the best we can do, right? Is there any way to improve on this without reopening questions that we felt we answered satisfactorily enough on Friday?

dhermes replied:
Your prose is hard to parse :grinning: 

We discussed two options:
- A single service account credentials class with factory constructors
- Three (or more) service account credentials classes

I preferred the first but said I couldn't justify it with a strong argument. I don't think there is any silver bullet here but think the fewer classes we have, the happier users will be.

We could eliminate the "shoot one-self in foot" problem by removing
- `private_key_pkcs8_pem` (specific to JSON)
- `private_key_pkcs12` (specific to P12)
- `private_key_password` (specific to JSON)
- `private_key_id` (specific to JSON, makes no difference when sent as `kid` in an assertion)
- `service_account_id` (specific to JSON, also **NEVER** used anywhere)

from the constructor and just relying on a `signer` passed to the constructor to encapsulate the relevant info (this was the original goal of the signer's anyhow). This mirrors the way that the `Verifier`/`Signer` classes are implemented.

nathanielmanistaatgoogle commented:
As for what I think about making `from_json_keyfile` accept one of three objects as its first argument: `ServiceAccountCredentials.from_json_keyfile` is a wholly original method that neither implements nor overrides anything else, right? Does it need to be able to take the other two kinds of input or is it simply enough to provide some way to accept the other two kinds of input? Why not define `ServiceAccountCredentials.from_json_keydict` and `ServiceAccountCredentials.from_json_keyfilename` and then have all three internally and privately delegate to common code? Would that work?

dhermes commented:
I also wanted to discuss the fact that this implementation double-stores `scopes` and shadows lots of fields from the parent that are unused

Here is a quick rundown of everything in `__dict__` for a `ServiceAccountCredentials` object:

```
_kwargs -> {}
_private_key_id -> 9ad967...
_private_key_password -> None
_private_key_pkcs12 -> None
_private_key_pkcs8_pem -> -----BEGIN PRIVATE KEY-----...
_scopes -> https://www.googleapis.com/auth/userinfo.email
_service_account_email -> service-account@FOO.iam.gserviceaccount.com
_service_account_id -> 10639...
_signer -> <oauth2client._openssl_crypt.OpenSSLSigner object at 0x7f6d2801d090>
_user_agent -> None
access_token -> ya29.fQ...
assertion_type -> None
client_id -> None
client_secret -> None
id_token -> None
invalid -> False
refresh_token -> None
revoke_uri -> https://accounts.google.com/o/oauth2/revoke
scopes -> set([])
store -> None
token_expiry -> 2016-02-03 22:34:22.926351
token_info_uri -> None
token_response -> {u'access_token': u'ya29.fQ...', u'token_type': u'Bearer', u'expires_in': 3600}
token_uri -> https://accounts.google.com/o/oauth2/token
user_agent -> None
```

dhermes commented:
@nathanielmanistaatgoogle I'm fine with have three factories (`from_json_keyfile`, `from_json_keydict` and `from_json_keyfile_name`) for varying stages of parsed-ness. Is the reason for this just an aversion to allowing multiple types in a function signature?

dhermes committed:
Addressing review comments.

dhermes commented:
@nathanielmanistaatgoogle We should also discuss how to break this up. I think the first two PRs could be

- pure-Python module as a stand-alone
- make `service_account.py` use `oauth2client.crypt` instead of the module's own hand-rolled auth in current use
- remove `client.SignedJWTCredentials` (will require temporarily disabling / breaking the system tests for `.p12` keys)
- everything else?

nathanielmanistaatgoogle commented on oauth2client/_pure_python_crypt.py:
```diff
@@ -0,0 +1,184 @@
+
+    @classmethod
+    def from_string(cls, key, password='notasecret'):
+        """Construct a Signer instance from a string.
```
Do you mean `an RsaSigner instance`?

nathanielmanistaatgoogle commented on oauth2client/_pure_python_crypt.py:
```diff
@@ -0,0 +1,184 @@
+                      files.
+
+        Returns:
+            Signer instance.
```
`RsaSigner instance` here too?

nathanielmanistaatgoogle commented on oauth2client/_pure_python_crypt.py:
```diff
@@ -0,0 +1,184 @@
+            Signer instance.
+
+        Raises:
+            NotImplementedError if the key isn't in PEM format.
```
`ValueError` is what is both what the code actually raises and what is appropriate to raise in this circumstance (no need to restate the larger philosophical difference here).

dhermes replied:
Copy-pasta strikes!

nathanielmanistaatgoogle commented on oauth2client/_pure_python_crypt.py:
```diff
@@ -0,0 +1,184 @@
+
+    @classmethod
+    def from_string(cls, key_pem, is_x509_cert):
+        """Construct a Verified instance from a string.
```
`an RsaVerifier instance`?

nathanielmanistaatgoogle commented on oauth2client/_pure_python_crypt.py:
```diff
@@ -0,0 +1,184 @@
+                          is expected to be an RSA key in PEM format.
+
+        Returns:
+            Verifier instance.
```
`RsaVerifier instance`?

nathanielmanistaatgoogle commented:
Responding to "I also wanted to discuss the fact that this implementation double-stores scopes and shadows lots of fields from the parent that are unused": my first reaction is "grumble grumble public classes grumble grumble implementation inheritance grumble grumble cuss words cuss words", but that's not productive. So my second reaction is: that's just kind of a thing that happens with deep inheritance hierarchies, but does any of it negatively effect either the implementation complexity or the correctness of the particular leaf class currently under review?

nathanielmanistaatgoogle commented:
Responding to "Is the reason for this just an aversion to allowing multiple types in a function signature?": yes. In my experience ad-hoc polymorphism always harms the long-term maintenance of a codebase and should be avoided wherever it can be avoided.

nathanielmanistaatgoogle commented:
Responding to "We should also discuss how to break this up": that does sound like a good plan. Certainly "introduce new code in its own commit" is a winning first step. Why break the system tests? Would it just be a matter of changing the out-of-repository artifacts of those tests at the same time as changing the in-repository artifacts of those tests?

dhermes committed:
Addressing review comments.

dhermes commented:
@nathanielmanistaatgoogle Pushed a new commit to address the docstring issues and to remove `_token_uri` and `_revoke_uri` from the svc. acct. creds constructor.

RE: Breaking system tests, I just meant that if we have a commit where `client.SignedJWTCredentials` is deleted and `ServiceAccountCredentials` doesn't support `.p12` keys, then the system test corresponding to a `.p12` key can't be run.

nathanielmanistaatgoogle commented:
Responding to: "if we have a commit where client.SignedJWTCredentials is deleted and ServiceAccountCredentials doesn't support .p12 keys, then the system test corresponding to a .p12 key can't be run" - that would be an _intermediate_ commit, right? ... and we don't even care if tests pass at _intermediate_ commit points, only at head-of-pull-request and head-of-repository, right?

dhermes commented:
Well my plan was for 4 PRs, not 4 commits. You can break it up as you like, not worth discussing with 4 posts (which we've done already).

nathanielmanistaatgoogle commented:
Ah, okay. I have no objection to four pull requests if that's your preference. Thank you for the clarity.

nathanielmanistaatgoogle commented on oauth2client/client.py:
```diff
@@ -1230,17 +1233,17 @@ def create_scoped(self, scopes):
-        # this non-top-level import.
-        from oauth2client.service_account import _ServiceAccountCredentials
-        data = json.loads(_from_bytes(s))
+        #                  this non-top-level import.
```
Did a formatting tool mutate this line? Not that it matters, just curious.

nathanielmanistaatgoogle commented on oauth2client/service_account.py:
```diff
@@ -12,96 +12,281 @@
+            else:
+                raise KeyError('JSON keyfile missing key', key)
+
+        return cls(**constructor_kwargs)
```
Normally I say "always pass positional arguments positionally and keyword arguments by keyword", but this expression looks like it will pass `service_account_email` by keyword. Have I gotten that right? No need to change anything, I'm just making sure I understand.

nathanielmanistaatgoogle commented on oauth2client/service_account.py:
```diff
@@ -12,96 +12,281 @@
+        super(ServiceAccountCredentials, self).__init__(
+            None, user_agent=user_agent)
 
-        self._service_account_id = service_account_id
```
This is just going away? Okay. What was it?

Oh, wait, no, it just moved down.

dhermes committed:
Fixing create_scoped to set token|revoke_uri after the fact.

dhermes commented:
Closing in lieu of #395, #396, #397, #398, #399
